src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -316,23 +310,34 @@ void MergeTreeBackgroundExecutor<Queue>::routine(TaskRuntimeDataPtr item)
         ALLOW_ALLOCATIONS_IN_SCOPE;
         query_id = item->task->getQueryId();
         need_execute_again = item->task->executeStep();
+
+        if (!need_execute_again)
+        {
+            complete_task(std::move(item));
+            return;
+        }
     }
     catch (...)
     {
         if (item->task->printExecutionException())
             printExceptionWithRespectToAbort(log, query_id);
+
+        try
+        {
+            ALLOW_ALLOCATIONS_IN_SCOPE;
+            item->task->cancel();
```

Comment:
Right now `auto cancel_task =...`
that lambda does not actually cancel task. It is canceled in that line.

How about put that code on the lambda but before the lock?

Reply:
> Right now auto cancel_task =... that lambda does not actually cancel task. It is canceled in that line.

Yes. Is this a problem? It's not a generic function that can be reused, and there's only one place where we call `cancel_task` so maybe we can remove this lambda at all 

> How about put that code on the lambda but before the lock?

We could do that, but then we would have to put exactly the same code in `complete_task` too